File: rust/2024/01/2/src/main.rs
use std::{
    collections::HashMap,
    fs::File,
    io::{BufRead, BufReader, Read},
    time::Instant,
};

fn main() {
    let start = Instant::now();

    let file = File::open("input.txt").expect("opening input file");
    println!("{}", process(file));

    println!("{:?}", start.elapsed());
}

fn process(file: impl Read) -> u32 {
    let mut locations_1 = Vec::new();
    let mut locations_2 = HashMap::new();

    for line in BufReader::new(file).lines().map_while(Result::ok) {
        let mut locations = line.split_whitespace();

        let location_1: u32 = locations.next().unwrap().parse().unwrap();
        locations_1.push(location_1);

        let location_2: u32 = locations.next().unwrap().parse().unwrap();
        let location_2_value = locations_2
            .get(&location_2)
            .map_or(1, |old_value| old_value + 1);
        locations_2.insert(location_2, location_2_value);
    }

    locations_1
        .into_iter()
        .map(|location_1| location_1 * locations_2.get(&location_1).unwrap_or(&0))
        .sum()
}

#[cfg(test)]
mod tests {
    use crate::process;
    use std::io::BufReader;

    #[test]
    fn test_example_result() {
        let lines = BufReader::new(
            "3   4
4   3
2   5
1   3
3   9
3   3"
                .as_bytes(),
        );

        assert_eq!(process(lines), 31);
    }
}

File: zig/2024/01/2/main.zig
const std = @import("std");
const file: []const u8 = @embedFile("input.txt");

pub fn main() !void {
    const result = try process(file);

    std.debug.print("{}\n", .{result});
}

fn process(input: []const u8) !u32 {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    var locations_1 = std.ArrayList(u32).init(allocator);
    var locations_2 = std.ArrayList(u32).init(allocator);

    var lines = std.mem.tokenizeScalar(u8, input, '\n');
    while (lines.next()) |line| {
        var locations = std.mem.tokenizeScalar(u8, line, ' ');

        const location_1 = try std.fmt.parseInt(u32, locations.next().?, 10);
        try locations_1.append(location_1);

        const location_2 = try std.fmt.parseInt(u32, locations.next().?, 10);
        try locations_2.append(location_2);
    }

    std.mem.sort(u32, locations_1.items, {}, std.sort.asc(u32));
    std.mem.sort(u32, locations_2.items, {}, std.sort.asc(u32));

    var result: u32 = 0;
    for (locations_1.items) |location_1| {
        var number_of_match: u32 = 0;
        for (locations_2.items) |location_2| {
            if (location_1 == location_2) {
                number_of_match += 1;
            }
        }
        result += location_1 * number_of_match;
    }

    return result;
}

test "process with example" {
    const result = process(
        \\3   4
        \\4   3
        \\2   5
        \\1   3
        \\3   9
        \\3   3
    );

    try std.testing.expectEqual(31, result);
}
